File: Lesson29_Joystick_FlowingLeds/code/python/Lesson29_Joystick_FlowingLeds.py
from gpiozero import LED
import smbus
import time

PCF8591_ADDR = 0x48
Y_CH = 0x41
X_CH = 0x42
Z_CH = 0x43

led_pins = [17,18,27,22,23]
leds = [LED(pin,active_high=False) for pin in led_pins]

bus = smbus.SMBus(1)

def read_adc(channel):
    try:
        bus.write_byte(PCF8591_ADDR,channel)
        bus.read_byte(PCF8591_ADDR)
        return bus.read_byte(PCF8591_ADDR)
    except Exception:
        return -1

def led_on(n):
    leds[n].on()

def led_off(n):
    leds[n].off()

def led_forward():
    for i in range(len(leds)):
        led_on(i)
        time.sleep(0.3)
        led_off(i)

def led_backward():
    for i in range(len(leds)-1,-1,-1):
        led_on(i)
        time.sleep(0.3)
        led_off(i)

try:
    while True:
        x = read_adc(X_CH)
        y = read_adc(Y_CH)
        z = read_adc(Z_CH)

        if x < 0 or y < 0 or z < 0:
            print("I2C read error")
            time.sleep(0.1)
            continue

        if x < 50:
            print(f"left X={x} Y={y} Z={z}")
            led_backward()
        elif x > 220:
            print(f"right X={x} Y={y} Z={z}")
            led_forward()
        elif y < 100:
            print(f"up X={x} Y={y} Z={z}")
            led_forward()
        elif y > 230:
            print(f"down X={x} Y={y} Z={z}")
            led_backward()

        time.sleep(0.1)

except KeyboardInterrupt:
    pass

finally:
    for led in leds:
        led.off()
    bus.close()

File: Lesson29_Joystick_FlowingLeds/code/c/Lesson29_Joystick_FlowingLeds.c
#include <lgpio.h>
#include <stdio.h>

#define PCF8591_ADDR 0x48
#define Y_CH 0x41
#define X_CH 0x42
#define Z_CH 0x43

const int ledPins[] = {17,18,27,22,23};
#define LED_COUNT 5

int readADC(int fd,int channel)
{
    int value;
    if(lgI2cWriteByte(fd,channel)<0)
        return -1;
    if(lgI2cReadByte(fd)<0)
        return -1;
    value=lgI2cReadByte(fd);
    return value;
}

void led_on(int h,int n)
{
    lgGpioWrite(h,ledPins[n],0);
}

void led_off(int h,int n)
{
    lgGpioWrite(h,ledPins[n],1);
}

void led_forward(int h)
{
    int i;
    for(i=0;i<LED_COUNT;i++){
        led_on(h,i);
        lguSleep(0.3);
        led_off(h,i);
    }
}

void led_backward(int h)
{
    int i;
    for(i=LED_COUNT-1;i>=0;i--){
        led_on(h,i);
        lguSleep(0.3);
        led_off(h,i);
    }
}

int main(void)
{
    int h=lgGpiochipOpen(4);
    if(h<0){
        printf("Failed to open gpiochip4, try gpiochip0\n");
        h=lgGpiochipOpen(0);
        if(h<0)
            return -1;
    }

    int fd=lgI2cOpen(1,PCF8591_ADDR,0);
    if(fd<0){
        printf("Failed to open I2C device\n");
        lgGpiochipClose(h);
        return -1;
    }

    int x,y,z;
    int i;

    for(i=0;i<LED_COUNT;i++)
        lgGpioClaimOutput(h,0,ledPins[i],1);

    while(1){
        x=readADC(fd,X_CH);
        y=readADC(fd,Y_CH);
        z=readADC(fd,Z_CH);

        if(x<0||y<0||z<0){
            printf("I2C read error\n");
            lguSleep(0.1);
            continue;
        }

        if(x<50){
            printf("left X=%d Y=%d Z=%d\n",x,y,z);
            led_backward(h);
        }
        else if(x>220){
            printf("right X=%d Y=%d Z=%d\n",x,y,z);
            led_forward(h);
        }
        else if(y<100){
            printf("up X=%d Y=%d Z=%d\n",x,y,z);
            led_forward(h);
        }
        else if(y>230){
            printf("down X=%d Y=%d Z=%d\n",x,y,z);
            led_backward(h);
        }

        lguSleep(0.1);
    }

    lgI2cClose(fd);
    lgGpiochipClose(h);
    return 0;
}